sklearn/ensemble/tests/test_gradient_boosting_loss_functions.py

```diff
@@ -34,23 +34,26 @@ def test_binomial_deviance():
     assert_almost_equal(bd(np.array([1.0, 0.0, 0.0]),
                            np.array([100.0, -100.0, -100.0])), 0)
 
-    # check if same results as alternative definition of deviance (from ESLII)
+    # check if same results as alternative definition of deviance, from ESLII
+    # Eq. (10.18): -loglike = log(1 + exp(-2*z*f))
+    # Note:
+    # - We use y = {0, 1}, ESL (10.18) uses z in {-1, 1}, hence y=2*y-1
+    # - ESL 2*f = pred_raw, hence the factor 2 of ESL disappears.
+    # - Deviance = -2*loglike + .., hence a factor of 2 in front.
     def alt_dev(y, pred):
-        return np.mean(np.logaddexp(0.0, -2.0 * (2.0 * y - 1) * pred))
+        return 2*np.mean(np.logaddexp(0.0, -(2.0 * y - 1) * pred))
```

Comment:
can we please:

- not use logaddexp (for readability)
- use raw_pred instead of pred
- explicitly define z as (2.0 * y - 1)
- not use `2.0` and just use `2`

Also I'm confused now... This expression in ESL is 
`deviance = log(1 + exp(-2 * z * raw_pred))`
but with the new changes we have
` 2* log(1 + exp(-z * raw_pred))`

Shouldn't the `alt_dev` stick to the definition of the book?

Reply:
1) Factor 2 in font of `mean`: ESL defines `-loglike`, while `deviance = -2*loglike + ...`.
2) Factor 2 in `exp`: ESL uses `f` for raw predictions that relates to `2 * f = raw_prediction`.